File: ex32/ex32.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {

    let n = s.length;
    let dp = new Array(n).fill(0);
    let maxans = 0;
    for (let i = 1; i < n; ++i) {
        if (s[i] == ')') {
            if (s[i - 1] == '(')
                dp[i] = (i - 2 >= 0 ? dp[i - 2] : 0) + 2;
            else if (i - dp[i - 1] - 1 >= 0 && s[i - dp[i - 1] - 1] == '(')
                dp[i] = dp[i - 1] + 2 + (i - dp[i - 1] - 2 >= 0 ? dp[i - dp[i - 1] - 2] : 0);
        }
        maxans = Math.max(maxans, dp[i]);

    }
    return maxans;


};
console.log(longestValidParentheses('(()))'))

File: ex32/ex32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        n = len(s)
        dp = [0] * n
        maxans = 0
        for i in range(1, n):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = (dp[i - 2] if i - 2 >= 0 else 0) + 2;
                elif i - dp[i - 1] - 1 >= 0 and s[i - dp[i - 1] - 1] == '(':
                    dp[i] = dp[i - 1] + 2 + (dp[i - dp[i - 1] -2] if i - dp[i - 1] - 2 >= 0 else 0);
            maxans = max(maxans, dp[i])
        return maxans